core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -349,7 +349,7 @@ class KafkaApis(val requestChannel: RequestChannel,
    * Handle a produce request
    */
   def handleProducerRequest(request: RequestChannel.Request) {
-    val produceRequest = request.body.asInstanceOf[ProduceRequest]
+    val produceRequest = request.takeBodyOwnership.asInstanceOf[ProduceRequest]
```

Comment:
Instead of doing this, could we not just pass the relevant fields to the callback as a top-level method (as suggested in the other comment)?